### ISSUE TITLE
Does a single AMD GPU can be shared among containers? Things like MPS of Nvidia

### ISSUE THREAD

**Issue by 676384417** [RadeonOpenCompute/ROCm-docker#62]
Does a single AMD GPU can be shared among several containers? 
Nvidia GPU has things like MPS https://docs.deep-hybrid-datacloud.eu/en/latest/technical/others/gpu-sharing-with-mps.html, but I'm not aware of any similar feature of AMD GPU.

**Comment by fxkamd** [comment]
The ROCm driver supports multiple concurrent processes on the GPU natively. We don't need MPS for that. This should also work for multiple containers.

The processes end up sharing GPU resources, which can lead to performance and functionality problems. Compute queues virtual memory IDs can be oversubscribed, but when that happens, performance drops significantly. GPU memory cannot be oversubscribed. This may result in out-of-memory errors causing memory-hungry applications to fail.

**Comment by 676384417** [comment]
> The ROCm driver supports multiple concurrent processes on the GPU natively. We don't need MPS for that. This should also work for multiple containers.
> 
> The processes end up sharing GPU resources, which can lead to performance and functionality problems. Compute queues virtual memory IDs can be oversubscribed, but when that happens, performance drops significantly. GPU memory cannot be oversubscribed. This may result in out-of-memory errors causing memory-hungry applications to fail.

Thank you!  Any guides on how to config ROCm-docker start several containers and share the same GPU device?

**Comment by fxkamd** [comment]
I don't think you need to do anything special. Just start multiple containers. They will share the GPU in the same way as multiple processes in the same container.

**Comment by 676384417** [comment]
> I don't think you need to do anything special. Just start multiple containers. They will share the GPU in the same way as multiple processes in the same container.

But then how to do isolation? e.g limit the gpu resources usage if coming from a low priority container.  Do you aware any solution to do GPU performance QoS among different processes/containers?
Thanks!

**Comment by fxkamd** [comment]
> > I don't think you need to do anything special. Just start multiple containers. They will share the GPU in the same way as multiple processes in the same container.
> 
> But then how to do isolation? e.g limit the gpu resources usage if coming from a low priority container. Do you aware any solution to do GPU performance QoS among different processes/containers?
> Thanks!

Well, that's a different question. There is currently no support in the driver for QoS or resource isolation between containers. Some proposals for cgroup controls to manage GPU resources are being discussed upstream to solve this problem.
